pkg/azure/provider/provider.go

```diff
@@ -0,0 +1,200 @@
+	if err != nil {
+		return
+	}
+	subnet, err := helpers.GetSubnet(ctx, d.factory, connectConfig, providerSpec)
+	if err != nil {
+		return
+	}
+
+	nicID, err := helpers.CreateNICIfNotExists(ctx, d.factory, connectConfig, providerSpec, subnet, nicName)
+	if err != nil {
+		return
+	}
+
+	vm, err := helpers.CreateOrUpdateVM(ctx, d.factory, connectConfig, providerSpec, imageReference, plan, req.Secret, nicID, vmName)
+	if err != nil {
+		return
+	}
+	resp = helpers.ConstructCreateMachineResponse(providerSpec.Location, vmName)
+	helpers.LogVMCreation(providerSpec.Location, providerSpec.ResourceGroup, vm)
+	return
+}
+
+func (d defaultDriver) DeleteMachine(ctx context.Context, req *driver.DeleteMachineRequest) (resp *driver.DeleteMachineResponse, err error) {
+	defer instrument.RecordDriverAPIMetric(err, deleteMachineOperationLabel, time.Now())
+	providerSpec, connectConfig, err := helpers.ExtractProviderSpecAndConnectConfig(req.MachineClass, req.Secret)
+	if err != nil {
+		return
+	}
+	var (
+		resourceGroup = providerSpec.ResourceGroup
+		vmName        = strings.ToLower(req.Machine.Name)
+	)
+	// Check if Deletion of the machine (VM, NIC, Disks) can be completely skipped.
+	skipDelete, err := helpers.SkipDeleteMachine(ctx, d.factory, connectConfig, resourceGroup)
+	if err != nil {
+		return
+	}
+	if skipDelete {
+		klog.Warningf("skipping delete of Machine [ResourceGroup: %s, Name: %s] since the ResourceGroup no longer exists", resourceGroup, req.Machine.Name)
+		resp = &driver.DeleteMachineResponse{}
+		return
+	}
+
+	vmAccess, err := d.factory.GetVirtualMachinesAccess(connectConfig)
+	if err != nil {
+		err = status.Error(codes.Internal, fmt.Sprintf("failed to create virtual machine access to process request: [resourceGroup: %s, vmName: %s], Err: %v\n", resourceGroup, vmName, err))
+		return
+	}
+	vm, err := clienthelpers.GetVirtualMachine(ctx, vmAccess, resourceGroup, vmName)
+	if err != nil {
+		err = status.Error(codes.Internal, fmt.Sprintf("failed to get virtual machine for VM: [resourceGroup: %s, name: %s], Err: %v", resourceGroup, vmName, err))
+		return
+	}
+	/*
+		It is possible to have left over NIC's and Disks even if the VM is no longer there. This is made possible because in the earlier version of this provider
+		implementation the cascade-delete is not enabled for NICs and Disks on deletion of the VM. Thus, it's possible that while the VM gets deleted the NIC's and Disks are left behind.
+		Once all the VirtualMachines are launched with cascade-delete enabled for NICs and Disks then this can be removed.
+	*/
+	if vm == nil {
+		klog.Infof("VirtualMachine [resourceGroup: %s, name: %s] does not exist. Skipping deletion of VirtualMachine", providerSpec.ResourceGroup, vmName)
+		// check if there are leftover NICs and Disks that needs to be deleted.
+		if err = helpers.CheckAndDeleteLeftoverNICsAndDisks(ctx, d.factory, vmName, connectConfig, providerSpec); err != nil {
+			return
+		}
```

Comment:
could you also add a log here to specify that only NIC and Disks are deleted. Basically we want a way to differentiate b/w cascade delete and this delete.

Reply:
added. PTAL